### PR TITLE
B ext

### PR DESCRIPTION
**Pull request by jerryz123**
A simpler approach than the previous attempt.

**Related issue**: 

**Type of change**: bug report | feature request | other enhancement

**Impact**: no functional change | API addition (no impact on existing code) | API modification

**Development Phase**: proposal |  implementation

**Release Notes**

### CODE REVIEW

**Comment by sequencer** [comment]
omg

File: src/main/scala/rocket/ALU.scala [hunk 1]
```diff
@@ -4,13 +4,14 @@
...
-import chisel3.util.{BitPat, Fill, Cat, Reverse}
+import chisel3.util.{BitPat, Fill, Cat, Reverse, PriorityEncoderOH, PopCount, MuxLookup}
 import org.chipsalliance.cde.config.Parameters
 import freechips.rocketchip.tile.CoreModule
+import freechips.rocketchip.util._
 
-class ALUFN {
-  val SZ_ALU_FN = 4
-  def FN_X    = BitPat("b????")
+object ALU {
```
Comment (sequencer) [inline src/main/scala/rocket/ALU.scala:12]: So question is here: the uop of ALU goes from 4 width to 5 width, a little increasing for all cores, but anyway in current rocket-chip methodology, we don't need to take it seriously.

File: src/main/scala/rocket/ALU.scala [hunk 2]
```diff
@@ -41,21 +56,22 @@ class ALUFN {
 
   def isMulFN(fn: UInt, cmp: UInt) = fn(1,0) === cmp(1,0)
   def isSub(cmd: UInt) = cmd(3)
-  def isCmp(cmd: UInt) = cmd >= FN_SLT
+  def isCmp(cmd: UInt) = (cmd >= FN_SLT && cmd <= FN_SGEU)
+  def isMaxMin(cmd: UInt) = (cmd >= FN_MAX && cmd <= FN_MINU)
```
Comment (sequencer) [inline src/main/scala/rocket/ALU.scala:60]: I'm sightly nervous on these code, maybe we can believe synthesis code? But I personally think we can decode to a onehot signal at the decode stage(but the uop will increase). This is an affordable trade off.

File: src/main/scala/rocket/ALU.scala [hunk 3]
```diff
@@ -64,50 +80,97 @@ abstract class AbstractALU[T <: ALUFN](val aluFn: T)(implicit p: Parameters) ext
...
+  val in1_bytes = io.in1.asTypeOf(Vec(xLen / 8, UInt(8.W)))
+  val orcb = VecInit(in1_bytes.map(b => Fill(8, b =/= 0.U))).asUInt
+  val rev8 = VecInit(in1_bytes.reverse).asUInt
+  val unary = MuxLookup(io.in2(11,0), count)(Seq(
+    0x287.U -> orcb,
+    (if (xLen == 32) 0x698 else 0x6b8).U -> rev8,
+    0x080.U -> io.in1(15,0),
+    0x604.U -> Fill(xLen-8, io.in1(7)) ## io.in1(7,0),
+    0x605.U -> Fill(xLen-16, io.in1(15)) ## io.in1(15,0)
+  ))
```
Comment (sequencer) [inline src/main/scala/rocket/ALU.scala:149]: RISC-V is elegant? :melting_face: 
For other future note, this is defined https://github.com/riscv/riscv-opcodes/blob/07b21cc5143a15959eda12e30aa40cea0971efe0/rv_zbb#L4-L6
Used for zvbb...

This remains some discussion on the where should unary decoding be? at DEC stage or at EXE stage?
For me I'd like to put it to dec, but the rocket-chip dec is hard to maintain...

File: src/main/scala/rocket/ALU.scala (hunk 1 above)
Comment (jerryz123) [reply src/main/scala/rocket/ALU.scala:12]: The high bit is useless when no B ext is enabled. It should be optimized away at synthesis

File: src/main/scala/rocket/ALU.scala (hunk 2 above)
Comment (jerryz123) [reply src/main/scala/rocket/ALU.scala:60]: This is only a handful of gates on `io.fn`, its just a bitmask ignoring the lower 2 bits of `io.fn`.

File: src/main/scala/rocket/ALU.scala (hunk 3 above)
Comment (jerryz123) [reply src/main/scala/rocket/ALU.scala:149]: This should be done at DEC really, but I don't think it makes a big difference in the grand scheme

File: overlay.nix [hunk 4]
```diff
@@ -8,7 +8,7 @@ final: prev: {
   });
   riscvTests = final.pkgsCross.riscv64-embedded.stdenv.mkDerivation rec {
     pname = "riscv-tests";
-    version = "55bbcc8c06637a31cc01970881ba8072838a9121";
+    version = "f2f748ebb9cf8ea049103f85c4cbf7e8a2927b16";
```
Comment (sequencer) [inline overlay.nix:11]: Forgot to update sha256 to `sha256-E3RfrP+PFIYy9c/pY04jYPxeGpnfgWwjV8iwL5+s+9w=`

File: src/main/scala/rocket/ALU.scala (hunk 2 above)
Comment (cyyself) [reply src/main/scala/rocket/ALU.scala:60]: I think using `cmd === FN_MAXMIN` at `isMaxMin` will make it easier to maintain. We can also apply this to `isCmp`.

Comment (jerryz123) [reply src/main/scala/rocket/ALU.scala:60]: You mean use `BitPat` to do this?

Comment (cyyself) [reply src/main/scala/rocket/ALU.scala:60]: > You mean use `BitPat` to do this?

Yes.

Comment (sequencer) [reply src/main/scala/rocket/ALU.scala:60]: In the future, with our new Rocket framework, we'd like to use our new [rvdecoderdb](https://github.com/chipsalliance/rvdecoderdb) infra for this to handle the control signal generation, and they can be used for multiple-stage uop decoding.

**Comment by sequencer** [comment]
I saw CI passed in [here](https://github.com/chipsalliance/rocket-chip/actions/runs/10357436845/job/28669468965?pr=3667), let's get it into upstream!

**Review by sequencer — APPROVED** [review]

**Pull request merged**